pkg/gateway/gateway.go

```diff
@@ -228,3 +224,133 @@ func (gw *Gateway) GetNetwork(name string) (*Network, error) {
 func (gw *Gateway) Close() {
 	// future use
 }
+
+func createGatewayConfigProvider(config core.ConfigProvider, gw *Gateway) func() ([]core.ConfigBackend, error) {
+	return func() ([]core.ConfigBackend, error) {
+		configBackend, err := config()
+		if err != nil {
+			return nil, err
+		}
+		if len(configBackend) != 1 {
+			return nil, errors.New("invalid config file")
+		}
+
+		cfg := configBackend[0]
+
+		lhConfig := make([]core.ConfigBackend, 0)
+		lhConfig = append(lhConfig, createGatewayConfig(cfg, gw))
+
+		return lhConfig, nil
+	}
+}
+
+func createGatewayConfig(backend core.ConfigBackend, gw *Gateway) *gatewayConfig {
+	var matchers map[string][]map[string]string
+	if strings.ToUpper(os.Getenv("DISCOVERY_AS_LOCALHOST")) == "TRUE" {
+		matchers = createLocalhostMappings()
+	}
+
+	var channelConfig map[string]map[string]map[string]map[string]bool
+	_, exists := backend.Lookup("channels")
+	if !exists {
+		channelConfig = createDefaultChannelConfig(backend, gw)
+	}
+
+	return &gatewayConfig{
+		backend:    backend,
+		matchers:   matchers,
+		channelDef: channelConfig,
+	}
+}
+
+/* dynamically add the following to CCP:
+
+entityMatchers:
+  peer:
+    - pattern: ([^:]+):(\\d+)
+      urlSubstitutionExp: localhost:${2}
+      sslTargetOverrideUrlSubstitutionExp: ${1}
+      mappedHost: ${1}
+  peer:
+    - pattern: ([^:]+):(\\d+)
+      urlSubstitutionExp: localhost:${2}
+      sslTargetOverrideUrlSubstitutionExp: localhost
+      mappedHost: ${1}
+*/
+func createLocalhostMappings() map[string][]map[string]string {
+	matchers := make(map[string][]map[string]string)
+	peerMappings := make([]map[string]string, 0)
+	ordererMappings := make([]map[string]string, 0)
+	mappedHost := "${1}"
+
+	peerMapping := make(map[string]string)
+	peerMapping["pattern"] = "([^:]+):(\\d+)"
+	peerMapping["urlSubstitutionExp"] = "localhost:${2}"
+	peerMapping["sslTargetOverrideUrlSubstitutionExp"] = mappedHost
+	peerMapping["mappedHost"] = mappedHost
+	peerMappings = append(peerMappings, peerMapping)
+
+	matchers["peer"] = peerMappings
+
+	ordererMapping := make(map[string]string)
+	ordererMapping["pattern"] = "([^:]+):(\\d+)"
+	ordererMapping["urlSubstitutionExp"] = "localhost:${2}"
+	ordererMapping["sslTargetOverrideUrlSubstitutionExp"] = "localhost"
+	ordererMapping["mappedHost"] = mappedHost
+	ordererMappings = append(ordererMappings, ordererMapping)
+
+	matchers["orderer"] = ordererMappings
+	return matchers
+}
+
+/* dynamically add the following to CCP:
+
+channels:
+  _default:
+    peers:
+      <gateway_peer_name>:
+        endorsingPeer: true
+        chaincodeQuery: true
+        ledgerQuery: true
+        eventSource: true
+*/
+func createDefaultChannelConfig(backend core.ConfigBackend, gw *Gateway) map[string]map[string]map[string]map[string]bool {
```

Comment:
it would be clearer to pass an org argument rather than the entire gateway object. i was wondering why gw was needed a couple calls up the stack :).

Reply:
Yes, the reason I was passing the gateway was because the org member is not assigned until a bit later.  But you're right, so I've changed it so it now passes a function that returns the org, which gets invoked when needed. Thanks